R/vcf2bedpe.R

```diff
@@ -0,0 +1,163 @@
+                               STRAND_B = rep('_', length(simple.bp$CHR2)),
+                               SVTYPE = simple.bp$SVTYPE
+                               );
+  } else {
+    # assign empty bedpe
+    simple.bedpe <- bedpe.cols;
+  }
+  # Convert paired breakends with MATEID
+  bnd.bp <- subset(x, SVTYPE == 'BND');
+  if (nrow(bnd.bp > 0)) {
+    catv('PROCESSING BND BREAKENDS\n')
+    rownames(bnd.bp) <- bnd.bp$ID;
+    if ('MATEID' %in% names(bnd.bp)) {
+      bnd.bp <- subset(bnd.bp, !is.na(MATEID));
+      name <- get.name(bnd.bp);
+      bnd_pair <- list();
+      for (id in rownames(bnd.bp)) {
+        if (!id %in% bnd_pair) {
+          bnd_pair[[id]] <- bnd.bp[id, 'MATEID'];
+        }
+      }
+      var <- bnd.bp[names(bnd_pair), ]; 
+      mates <- bnd.bp[unlist(unname(bnd_pair)),];
+      if (!'STRAND' %in% names(bnd.bp) & 'MATESTRAND' %in% names(bnd.bp)) {
+        # for delly v0.7.8 because the INFO:MATESTRAND is read while the INFO:STRAND is not 
+        # STRAND is not in the header so it is not read by bedr
+        var$STRAND <- mates$MATESTRAND;
+        mates$STRAND <- var$MATESTRAND;
+      }
+      coordsA <- adjust.coordinates(var, 'CIPOS', var$POS,  var$POS);
+      coordsB <- adjust.coordinates(mates, 'CIPOS', mates$POS, mates$POS);
+      strandA <- get.strand(var);
+      strandB <- get.strand(mates);
+      svtype <- ifelse(!is.null(var$SIMPLE_TYPE), var$SIMPLE_TYPE, 'BND');
+      bnd.bedpe <- data.frame(CHROM_A = var$CHROM,
+                              START_A = coordsA$start,
+                              END_A = coordsA$end,
+                              CHROM_B = mates$CHROM,
+                              START_B = coordsB$start,
+                              END_B = coordsB$end,
+                              ID = name,
+                              QUAL = var$QUAL,
+                              STRAND_A = strandA,
+                              STRAND_B = strandB,
+                              SVTYPE = svtype
+                              );
+      } else {
+        stop('MATEID is not present in the VCF file INFO field');
+        }
+  } else {
+    bnd.bedpe <- bedpe.cols;
+  }
+  bedpe_df <- rbind(simple.bedpe, bnd.bedpe);
+  bedpe_df <- bedpe_df[with(bedpe_df, order(CHROM_A, START_A)), ];
+  if (!is.null(filename)) {
+    write.table(bedpe_df,
+                file = filename,
+                col.names = header,
+                row.names = FALSE,
+                sep = '\t',
+                quote = FALSE
+                );
+  }
+  return(bedpe_df);
+}
+
+# get STRAND from INFO or ALT field
+get.strand <- function(df) {
+  strand <- rep('+', nrow(df));
+  if ('STRAND' %in% names(df)) { 
+    strand <- df$STRAND;
+  } else if (any(grepl('\\]|\\[', df$ALT))) {
+    strand[grepl('\\[.*\\[', df$ALT)] <- '-';
+  } else {
+    catv('STRAND information is not recorded in INFO:STRAND or ALT field. Returning default: + ')
+  }
+  return(strand);
+} 
+
```

Comment:
Did you want to move these helper functions to independent .R files (with their own .Rd)